runtime/gc_glue_java/ArrayletObjectModel.hpp

```diff
@@ -831,6 +831,160 @@ class GC_ArrayletObjectModel : public GC_ArrayletObjectModelBase
+	}
+
+	/**
+	 * Gets data pointer of a discontiguous indexable object.
+	 * Helper to get dataAddr field of discontiguous indexable objects.
+	 *
+	 * @return Pointer which points to discontiguous indexable object data
+	 */
+	MMINLINE void **
+	dataAddrSlotForDiscontiguous(J9IndexableObject *arrayPtr)
+	{
+		/* If double mapping is enabled only, arraylet will have a discontiguous layout.
+		 * If sparse-heap is enabled, arraylet will have a contiguous layout. For now
+		 * we can't simply Assert only the discontiguous case because there could also
+		 * exist hybrid arraylets (which will be dicontinued in the future) */
+		bool const compressed = compressObjectReferences();
+		void **dataAddrPtr = NULL;
+		if (compressed) {
+			dataAddrPtr = &((J9IndexableObjectDiscontiguousCompressed *)arrayPtr)->dataAddr;
+		} else {
+			dataAddrPtr = &((J9IndexableObjectDiscontiguousFull *)arrayPtr)->dataAddr;
+		}
+		return dataAddrPtr;
+	}
+
+	/**
+	 * Sets data pointer of a contiguous indexable object.
+	 * Sets the data pointer of a contiguous indexable object; in this case
+	 * dataAddr will point directly into the data right after dataAddr field
+	 * (data resides in heap).
+	 *
+	 * @param arrayPtr      Pointer to the indexable object whose size is required
+	 */
+	MMINLINE void
+	setDataAddrForContiguous(J9IndexableObject *arrayPtr)
+	{
+		void **dataAddrPtr = dataAddrSlotForContiguous(arrayPtr);
+		void *dataAddr = (void *)((U_64)arrayPtr + contiguousHeaderSize());
+		*dataAddrPtr = dataAddr;
+	}
+
+	/**
+	 * Sets data pointer of a discontiguous indexable object.
+	 * Sets the data pointer of a discontiguous indexable object; in this case
+	 * dataAddr will point to the contiguous representation of the data
+	 * which resides outside the heap (assuming double map/sparse-heap is enabled).
+	 * In case double map is disabled, the dataAddr will point to the first arrayoid
+	 * of the discontiguous indexable object, which also resides right after dataAddr
+	 * field.
+	 *
+	 * @param arrayPtr      Pointer to the indexable object whose size is required
+	 * @param dataAddr      Pointer which points to indexable object data
+	 */
+	MMINLINE void
+	setDataAddrForDiscontiguous(J9IndexableObject *arrayPtr, void *address)
+	{
+		/* If double mapping is enabled only, arraylet will have a discontiguous layout.
+		 * If sparse-heap is enabled, arraylet will have a contiguous layout. For now
+		 * we can't simply Assert only the discontiguous case because there could also
+		 * exist hybrid arraylets (which will be dicontinued in the future) */
+		void *calculatedDataAddr = address;
+		void **dataAddrPtr = dataAddrSlotForDiscontiguous(arrayPtr);
+
+		/* If calculatedDataAddr is NULL then we make dataAddr point to the first arrayoid */
+		/* Later on, when sparse-heap is enabled by default, we must assert dataAddr is not NULL */
+		if (NULL == calculatedDataAddr) {
+			calculatedDataAddr = (void *)((U_64)arrayPtr + discontiguousHeaderSize());
```

Comment:
it's more generic/optimal (in potential case this gets used on 32 bit) to cast to UDATA (what also discontiguousHeaderSize() returns).

same for contiguous